src/Core/Content/Product/SalesChannel/Listing/Processor/PagingListingProcessor.php

```diff
@@ -48,19 +48,32 @@ public function process(Request $request, ProductListingResult $result, SalesCha
             $result->setPage($page);
         }
 
-        $limit = $result->getCriteria()->getLimit() ?? $this->getLimit($result->getCriteria(), $context);
+        $limit = $result->getCriteria()->getLimit() ?? $this->getLimit($result->getCriteria(), $context, $request);
         $result->setLimit($limit);
     }
 
-    private function getLimit(Criteria $criteria, SalesChannelContext $context): int
+    private function getLimit(Criteria $criteria, SalesChannelContext $context, Request $request): int
     {
+        $defaultLimit = $this->config->getInt('core.listing.productsPerPage', $context->getSalesChannelId());
+        $requestLimit = $request->query->has('limit') ? $request->query->getInt('limit') : null;
```

Comment:
what about limit as request parameter? 

recently we removed setting the limit via request: https://github.com/shopware/shopware/pull/5831

we need to make sure, that only allowed limits are considered. therefore it would be nice, if you could test all cases that "codecov" commented on.

Reply:
On top of that, if you apply the limit here, doesn't it apply globally for every other slots?

I'd validate and apply the limit in the ProductListingCmsElementResolver so the limit is only applied for specific slot

Maybe something like this in ProductListingCmsElementResolver::enrich

```
        $slots = $resolverContext->getRequest()->query->get('slots') ?? [];
        if (\is_string($slots)) {
            $slots = explode('|', $slots);
        }
        
        if (in_array($slot->getId(), $slots) && $request->query->has('limit')) {
            $this->validateLimit($request, $slot);
            $criteria->setLimit($limit);
        }
```